# Chatgpt4o:

## Prompt 

Write a Python program for a basic calculator that supports addition, subtraction, multiplication, and division.

## Code-Analyse

Das folgende Python-Programm implementiert einen einfachen Taschenrechner mit vier Grundrechenarten: Addition, Subtraktion, Multiplikation und Division. Der Nutzer wird interaktiv aufgefordert, eine Rechenoperation sowie zwei numerische Werte (x und y) einzugeben. Anschließend wird das Ergebnis basierend auf der gewählten Operation berechnet und ausgegeben.

In [1]:
def add(x, y):
    return x + y

def subtract(x, y):
    return x - y

def multiply(x, y):
    return x * y

def divide(x, y):
    if y == 0:
        return "Error! Division by zero."
    return x / y


Die main()-Funktion übernimmt die Steuerung des Programmablaufs:

- Sie zeigt ein Menü an,

- liest die Benutzerwahl und Eingabewerte ein,

- ruft die passende Funktion auf und

- gibt das Ergebnis aus.

In [ ]:

def main():
    print("Basic Calculator")
    print("Select operation:")
    print("1. Add")
    print("2. Subtract")
    print("3. Multiply")
    print("4. Divide")

    choice = input("Enter choice (1/2/3/4): ")

    if choice in ['1', '2', '3', '4']:
        try:
            num1 = float(input("Enter first number: "))
            num2 = float(input("Enter second number: "))
        except ValueError:
            print("Invalid input! Please enter numeric values.")
            return

        if choice == '1':
            print(f"Result: {add(num1, num2)}")
        elif choice == '2':
            print(f"Result: {subtract(num1, num2)}")
        elif choice == '3':
            print(f"Result: {multiply(num1, num2)}")
        elif choice == '4':
            print(f"Result: {divide(num1, num2)}")
    else:
        print("Invalid choice!")

if __name__ == "__main__":
    main()



Der Code ist vollständig ausführbar und besitzt sogar eine kleine Fehlerbehandlung, wie beispielsweise bei der division durch null.

## Tests

### 1. Division durch 0

In [4]:
# Vorher den Codeblock mit den Operationen ausführen
print("1. Division durch 0")
print(divide(5, 0))  # Erwartet: "Error! Division by zero."

1. Division durch 0
Error! Division by zero.


### 2. Extrem große Zahlen

In [5]:
print(add(1e100, 1e100))        # Erwartet: 2e+100
print(subtract(1e100, 1e100))   # Erwartet: 0.0
print(multiply(1e100, 1e100))   # Erwartet: 1e+200
print(divide(1e100, 1e100))     # Erwartet: 1.0

2e+100
0.0
1e+200
1.0


### 3.Fließkommazahlen

In [6]:
print(add(3.5, 4.2))            # Erwartet: 7.7
print(subtract(3.5, 4.2))       # Erwartet: -0.7
print(multiply(3.5, 4.2))       # Erwartet: 14.7
print(divide(7.5, 2.5))         # Erwartet: 3.0


7.7
-0.7000000000000002
14.700000000000001
3.0


Die Werte -0.7000000000000002 und 14.700000000000001 entstehen durch die begrenzte Genauigkeit der Fließkommazahlen in der Computerarithmetik, bei der manche Dezimalzahlen nicht exakt dargestellt werden können, was zu minimalen Abweichungen führt.

### 4. Nähe zu Null

In [7]:
print(add(1e-10, 1e-10))        # Erwartet: 2e-10
print(subtract(1e-10, 1e-10))   # Erwartet: 0.0
print(multiply(1e-10, 1e10))    # Erwartet: 1.0
print(divide(1e-10, 1e10))      # Erwartet: 1e-20

2e-10
0.0
1.0
1.0000000000000001e-20


### 5.Negative Division

In [8]:
print(divide(-6, 3))            # Erwartet: -2.0
print(divide(6, -3))            # Erwartet: -2.0
print(divide(-6, -3))           # Erwartet: 2.0


-2.0
-2.0
2.0


### 6. Große Division (Vermeidung von Überlauffehlern)

In [9]:
print(divide(1e100, 1e50))      # Erwartet: 1e+50

9.999999999999999e+49


Das Ergebnis weicht minimal vom erwarteten Ergebnis ab.
Das liegt daran, dass Computer Dezimalzahlen intern in Binärform speichern. Viele Zahlen, die in Dezimalform exakt sind (wie 0.1, 1e100 oder 1e50), können im Binärsystem nicht exakt dargestellt werden. Das führt zu kleinen Rundungsfehlern bei der Berechnung. Würde man mathematisch 10^100 / 10^50 = 10^50 erwarten – also 1e+50. Aber intern speichert Python diese riesigen Zahlen in Gleitkommaform, und bei der Division tritt ein winziger Genauigkeitsverlust auf,

### 7. Falsche Argumente

In [10]:
try:
    print(add("abc", 5))        # Erwartet: TypeError
except TypeError as e:
    print("TypeError:", e)

TypeError: can only concatenate str (not "int") to str


TypeError wird nicht abgefangen